nevergrad/functions/powersystems/core.py

```diff
@@ -0,0 +1,154 @@
+        self.layers += [np.random.rand(layer_width, output_size)]
+        assert len(self.layers) == layers
+
+    def GetParamNumbers(self):
+        return sum([np.prod(l.shape) for l in self.layers])
+
+    def SetParams(self, ww):
+        w = [w for w in ww]
+        assert(len(w) == self.GetParamNumbers())
+        for l in self.layers:
+            s = np.prod(l.shape)
+            l = np.reshape(np.array(w[:s]), l.shape)
+            w = w[s:]
+
+    def GetOutput(self, inp):
+        output = np.array(inp).reshape(1, len(inp))
+        for l in [self.layers[i] for i in range(len(self.layers) - 1)]:
+            output = np.tanh(np.matmul(output, l))
+        return np.matmul(output, self.layers[-1])
+
+
+# Real life is more complicated! This is a very simple model.
+class PowerSystem(inst.InstrumentedFunction):
+    """
+    Parameters
+    ----------
+    nint intaum_stocks: number of stocks to be managed
+    depth: number of layers in the neural networks
+    width: number of neurons per hidden layer
+    """
+
+    def get_num_vars(self) -> List[Any]:
+        return [m.GetParamNumbers() for m in self.dam_managers]
+
+    def _simulate_power_system(self, input_x: np.ndarray) -> float:
+        dam_managers = self.dam_managers
+        x = list(input_x)
+        for a in dam_managers:
+            assert(len(x) >= a.GetParamNumbers())
+            a.SetParams(np.array(x[:a.GetParamNumbers()]))
+            x = x[a.GetParamNumbers():]
+        assert(len(x) == 0)
+        N = int(self.N)
+        # Assume empty initial stocks.
+        stocks = [0.] * N   
+        # Nonsense delays.
+        delay = [np.cos(i) for i in range(N)]
+        cost = 0.
+        # Loop on time steps.
+        consumption = 0.
+        hydro_prod_per_time_step: List[float] = []
+        consumption_per_time_step: List[float] = []
+        for t in range(int(365*24*self.number_of_years)):
+    
+            # Rain
+            for i in range(N):
+                stocks[i] += 0.5*(1.+np.cos(2*pi*t/(24*365) + delay[i])) * np.random.rand()
+            # Consumption model.
+            base_consumption = (self.constant_to_year_ratio*self.year_to_day_ratio 
+                    +0.5*self.year_to_day_ratio*(1.+np.cos(2*pi*t/(24*365))) + 0.5*(1.+np.cos(2*pi*t/24)))
+            if t == 0:
+                consumption = base_consumption
+            else:
+                consumption = max(0., consumption + self.consumption_noise*(np.random.rand()-.5) + self.back_to_normal * (base_consumption - consumption))
+            consumption_per_time_step += [consumption]
+            # Water usage.
+            needed = consumption
+    
+            # Setting inputs for all agents.
+            base_x = [np.cos(2*pi*t/24.), np.sin(2*pi*t/24.), np.cos(2*pi*t/(365*24)), np.sin(2*pi*t/(365*24)), needed, self.average_consumption, self.year_to_day_ratio, self.constant_to_year_ratio, self.back_to_normal, self.consumption_noise]
```

Comment:
Kind of a detail but math.cos is much faster than np.cos on scalars.

Reply:
Done, thx.